cache/utils.go

```diff
@@ -0,0 +1,152 @@
+// Utils provide a series of helper functions.
+
+package cache
+
+import (
+	"bytes"
+	"net/http"
+	"slices"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gophercloud/utils/openstack/clientconfig"
+	"github.com/openstack-exporter/openstack-exporter/exporters"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/expfmt"
+)
+
+// CollectCache collects the MetricsFamily for required clouds and services and stores in the cache.
+func CollectCache(
+	enableExporterFunc func(
+		string, string, string, []string, string, bool, bool, bool, bool, string, func() (string, error), log.Logger,
+	) (*exporters.OpenStackExporter, error),
+	multiCloud bool,
+	services map[string]*bool, prefix,
+	cloud string,
+	disabledMetrics []string,
+	endpointType string,
+	collectTime bool,
+	disableSlowMetrics bool,
+	disableDeprecatedMetrics bool,
+	disableCinderAgentUUID bool,
+	domainID string,
+	uuidGenFunc func() (string, error),
+	logger log.Logger,
+) error {
+	level.Info(logger).Log("msg", "Run collect cache job")
+	cacheBackend := GetCache()
+
+	clouds := []string{}
+
+	if multiCloud {
+		cloudsConfig, err := clientconfig.LoadCloudsYAML()
+		if err != nil {
+			return err
+		}
+		for cloud := range cloudsConfig {
+			clouds = append(clouds, cloud)
+		}
+	}
+	if cloud != "" && !multiCloud {
+		clouds = append(clouds, cloud)
+	}
+
+	enabledServices := []string{}
+	for service, disabled := range services {
+		if !*disabled {
+			enabledServices = append(enabledServices, service)
+		}
+	}
+
+	for _, cloud := range clouds {
+		level.Info(logger).Log("msg", "Start update cache data", "cloud", cloud)
+		// Update cloud's cache once finish all exporters' collection job. so we won't mix the old
+		// and new metrics in the cache and confuse users.
+		cloudCache := NewCloudCache()
+
+		for _, service := range enabledServices {
+			level.Info(logger).Log("msg", "Start collect cache data", "cloud", cloud, "service", service)
+			exp, err := enableExporterFunc(service, prefix, cloud, disabledMetrics, endpointType, collectTime, disableSlowMetrics, disableDeprecatedMetrics, disableCinderAgentUUID, domainID, nil, logger)
+			if err != nil {
+				// Log error and continue with enabling other exporters
+				level.Error(logger).Log("err", "enabling exporter for service failed", "cloud", cloud, "service", service, "error", err)
+				continue
+			}
+			registry := prometheus.NewPedanticRegistry()
```

Comment:
iiuc, this should be outside the inner for-loop, otherwise you create a registry for every service, and that `Gather` method only gather one service for each service in each cloud

Reply:
You can't do that here, let me explain why:

Imaging you put `prometheus.NewPedanticRegistry` in first for loop

```
for _, cloud := range clouds {
        registry := prometheus.NewPedanticRegistry()
        ...
        for _, service := range enabledServices {
                // If you have multiple services, the new exporter will append to the registry's exporter list and there is no way the flush it.
                // So the same metrics will run collect multiple times during the for loop iteration.
                registry.mustRegister(*exp)
                metricFamilies, err := registry.Gather()
                ...
        }
}
```

Reply:
The `Gather` should be called outside the inner `for-loop` though, i.e. when you finish registering all the exporter in the same cloud, at least that's how python prometheus client works, not sure if it's the same in Go

Reply:
OIkay thanks for the clarification, it seems a design decision to cache per service. Please consider this resolved